fvm/errors/errors_collector.go

```diff
@@ -0,0 +1,56 @@
+package errors
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+// Utility object for collecting Processor errors.
+type ErrorsCollector struct {
+	errors  error
+	failure error
+}
+
+func NewErrorsCollector() *ErrorsCollector {
+	return &ErrorsCollector{}
+}
+
+func (collector *ErrorsCollector) HasError() bool {
+	return collector.errors != nil || collector.failure != nil
+}
+
+func (collector *ErrorsCollector) ErrorOrNil() error {
+	if collector.failure == nil {
+		return collector.errors
+	}
+
+	if collector.errors == nil {
+		return collector.failure
+	}
+
+	return fmt.Errorf(
+		"%w (additional errors: %v)",
+		collector.failure,
+		collector.errors)
+}
+
+// This returns true if the processor should early exit (due to a fatal error).
+func (collector *ErrorsCollector) CollectShouldEarlyExit(err error) bool {
+	if err == nil {
+		return collector.failure != nil
+	}
+
+	if collector.failure != nil {
+		collector.errors = multierror.Append(collector.errors, err)
```

Comment:
does it mean that we only keep the first failure, and put all errors afterwards into `errors` even when it is a failure?

Reply:
we should early exit on failure in general so this case should never get triggered.  i added this just in case we forgot to do so (this enables us to track the secondary failure)